src/services/nodegit/NodeGitService.js

```diff
@@ -17,18 +18,142 @@ const DiffService = require('../diff/DiffService');
 
 class NodeGitService
 {
-    async createBlueprintsFolder(githubUrl)
+    /**
+     * Clones an existing repo (un/initialised) to tmp blueprints and creates a branch for new blueprint files
+     *
+     * @param {*} githubUrl
+     * @param {*} isBlueprintRepo
+     * @returns {
+     *              repo: {Git.Repository},
+     *              url: {String},
+     *              folderPath : {String},
+     *              index : {String},
+     *              branch : {String}
+     *          }
+     * @memberof NodeGitService
+     */
+    async createOrCloneBlueprintRepoFromGithubUrl(githubUrl, isBlueprintRepo)
+    {
+        const blueprintUrl = new Url(githubUrl)
+        let blueprintFolder;
+        let repo;
+
+        const signature = Git.Signature.now("Foo bar", "dom@kitset.io");
+
+        // -- Lets attempt to clone the repo if it is a blueprint repo. If it isn't then lets generate the blueprint repo name and attempt to clone that
+        if (!isBlueprintRepo)
+        {
+            blueprintUrl.pathname += "-blueprint";
+        }
+
+        const blueprintPathnameSplit = blueprintUrl.pathname.split("/");
+        const blueprintRepoName = blueprintPathnameSplit[blueprintPathnameSplit.length - 1];
+        const blueprintFolderPath = path.resolve(__dirname, ('../../../../sxd-git-projects/' + blueprintRepoName))
+
+        // -- First we need to ensure there is a repo that we can write to. We will not dynamically create a repo (maybe we can do this later but we will need to use the github API directly)
+        try
+        {
```

Comment:
To make testing easier id prob refactor steps like this block into their own functions with inputs and return statement expected